src/v/serde/serde.h

```diff
@@ -166,9 +177,64 @@ inline constexpr auto const are_bytes_and_string_different = !(
 template<typename T>
 void write(iobuf&, T);
 
+template<class R, class P>
+int64_t checked_duration_cast_to_nanoseconds(
+  const std::chrono::duration<R, P>& duration) {
+    static_assert(
+      __has_builtin(__builtin_mul_overflow),
+      "__builtin_mul_overflow not supported.");
+    using nano_period = std::chrono::nanoseconds::period;
+    using nano_rep = typename std::chrono::nanoseconds::rep;
+    // This is how duration_cast determines the output type.
+    using output_type = typename std::common_type<R, nano_rep, intmax_t>::type;
```

Comment:
>It will probably be easier to just say that output_type is intmax_t. And don't care of what R or nano_rep are. Well only ensure that R is a signed integer.

Right.. this was just an additional conservative  check, retained because there is no runtime overhead.

> In theory R can even be double and the code below may correctly convert it into integer nanoseconds, but that would require additional checks

Agree.. check the static assert in L428 (I should probably move it here?), removed support for floating point types for this reason (and perf).

> unless we build the code with libc++ or with -std=gnu++17.

Btw we do build with libc++.